pkg/controller/vsphere/reconciler.go

```diff
@@ -465,6 +511,66 @@ type virtualMachine struct {
 	Obj *object.VirtualMachine
 }
 
+func (vm *virtualMachine) getAncestors() ([]mo.ManagedEntity, error) {
+	client := vm.Obj.Client()
+	pc := client.ServiceContent.PropertyCollector
+
+	return mo.Ancestors(vm.Context, client, pc, vm.Ref)
+}
+
+// getRegionAndZone checks the virtual machine and each of its ancestors for the
+// given region and zone labels and returns their values if found.
+func (vm *virtualMachine) getRegionAndZone(c *rest.Client, regionLabel, zoneLabel string) (map[string]string, error) {
+	result := make(map[string]string)
+	tagsMgr := tags.NewManager(c)
+
+	objects, err := vm.getAncestors()
+	if err != nil {
+		klog.Errorf("Failed to get ancestors for %s: %v", vm.Ref, err)
+		return nil, err
+	}
+
+	for i := range objects {
+		obj := objects[len(objects)-1-i] // Reverse order.
+		klog.V(4).Infof("getRegionAndZone: Name: %s, Type: %s",
+			obj.Self.Value, obj.Self.Type)
+
+		tags, err := tagsMgr.ListAttachedTags(vm.Context, obj)
+		if err != nil {
+			klog.Warningf("Failed to list attached tags: %v", err)
+			return nil, err
+		}
+
+		for _, value := range tags {
+			tag, err := tagsMgr.GetTag(vm.Context, value)
+			if err != nil {
+				klog.Errorf("Failed to get tag: %v", err)
+				return nil, err
+			}
+
+			category, err := tagsMgr.GetCategory(vm.Context, tag.CategoryID)
+			if err != nil {
+				klog.Errorf("Failed to get tag category: %v", err)
+				return nil, err
+			}
+
+			switch {
+			case regionLabel != "" && category.Name == regionLabel:
+				result[regionKey] = tag.Name
+				klog.Infof("%s has region tag (%s) with value %s",
```

Comment:
I don't know how many ancestors a VM might have, but could this log be quite noisy? Similar with the one below? 
Looks like this codepath is expected to execute every reconcile, for each ancestor of the VM that has the right label right? Should this maybe be a more verbose log level?

Reply:
True. I've made it `V(2)` and now return early once we've found both tags, which is something it should have been doing anyway.